Introduce new Executor primitive incorporating samplomatic semantics

delapuente opened the pull request:
This RFC proposes the introduction of a new Executor primitive into Qiskit as a vendor-agnostic substrate for execution. The `Executor` provides three core components—`Executor`, `ExecutorProgram`, and `ExecutorResult`—that capture the relationship between a programming model engine, its semantics, and its results.

The immediate motivation is to make randomization semantics for error mitigation explicit and first-class in Qiskit. Today, such workflows are implicit and opaque; this RFC standardizes how users express their intent through annotations and portable representations. This enables reproducibility, fine-grained control, and composability, while letting providers implement and optimize their own expansion engines.

This RFC does not introduce a specific engine such as IBM’s Samplomatic into Qiskit. Instead, it defines the substrate and user-intent representation, leaving execution engines to providers. Follow-up RFCs may propose additional semantics, a reference `Executor`, or reimplementations of existing primitives (e.g., `Sampler` and `Estimator`) atop this foundation.

delapuente committed:
Create 0023-low-level-sampler.md

delapuente committed:
Update 0023-low-level-sampler.md

Drafting the initial justification.

delapuente committed:
Complete non-technical sections

delapuente committed:
Preliminar draft to implementing estimator on top of the low level sampler.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/RFCs?pullRequest=78) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/RFCs?pullRequest=78) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: delapuente<br/>:x: SALVADOR DE LA PUENTE<br/><hr/>**SALVADOR DE LA PUENTE** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/RFCs?pullRequest=78) it.</sub>

garrison commented on 0023-low-level-sampler.md:
```diff
@@ -0,0 +1,203 @@
+with circuit0.box([Twirl(), InjectNoise("my_noise")]):
+  circuit0.cx(0, 1)
+
+with circuit.box([Twirl(), BasisTransform(ref="my_basis")]):
+  circuit.measure_all()
```
I believe you must have meant

```suggestion
with circuit0.box([Twirl(), BasisTransform(ref="my_basis")]):
  circuit0.measure_all()
```

garrison commented on 0023-low-level-sampler.md:
```diff
@@ -0,0 +1,203 @@
+)
+
+sampling_program = QuantumProgram(shots=1024)
+sampling_program.append(
+  circuit0,
+
+)
```
It looks like some arguments are missing to this call to `.append()`.

garrison reviewed (commented):
:+1:

delapuente committed:
Draft for the new low-level primitive complete.

delapuente committed:
Update 0023-low-level-sampler.md

blakejohnson commented on 0023-low-level-sampler.md:
```diff
@@ -0,0 +1,188 @@
+offers two such primitives: the Sampler and the Estimator. The Estimator
+leverages advanced error mitigation techniques to enhance the quality of
+results from noisy quantum computers. On a noiseless quantum computer, the
+Estimator could instead be constructed atop the Sampler, simplifying its role.
```
The purpose of the Estimator extends beyond just mitigating noise. It also adds a layer abstraction around the concept of estimating observables, allowing implementors of Estimator interfaces to offer different methods for observable estimation (e.g. classical shadows) while preserving a common interface.

blakejohnson commented on 0023-low-level-sampler.md:
````diff
@@ -0,0 +1,188 @@
+the new low-level interface to implement a client-side basic estimator.
+
+```python
+from qiskit_ibm_runtime import QiskitRuntimeService, Session
````
```suggestion
from qiskit_ibm_runtime import QiskitRuntimeService, Session, NoiseLearner
```

blakejohnson reviewed (changes requested):
This RFC seems to introduce 3 concepts:
1. Sampling graphs (`samplex`)
2. `Executor` primitive
3. `QuantumProgram` data structure

However, (2) and (3) aren't described at all beyond their appearance in the example. (1) gets the most motivating text, but the RFC doesn't actually say enough about what it is beyond something that produces circuit variations.

In short, this looks rather incomplete at this stage.

delapuente committed:
Include NoiseLearner import in the example.

Co-authored-by: Blake Johnson <blakejohnson04@gmail.com>

delapuente committed:
Avoid name collision.

delapuente committed:
Update to disambiguating name for avoiding name collision.

delapuente committed:
Clarify the primitive definition.

delapuente committed:
Correct purpose of sending thousands/millions of circuit variations.

delapuente committed:
Update and rename 0023-low-level-sampler.md to 0023-executor-primitive.md

ihincks commented on 0023-executor-primitive.md:
```diff
@@ -0,0 +1,188 @@
+In Qiskit, the quantum computer interface is built around primitives, which are
+the fundamental operations carried by a quantum computer on a quantum circuit.
+Currently, Qiskit offers two such primitives: the Sampler and the Estimator.
+The Estimator leverages advanced error mitigation techniques to enhance the quality
+of results from noisy quantum computers. On a noiseless quantum computer, the
+Estimator could instead be constructed atop the Sampler, simplifying its role.
+However, given current noise levels, robust error mitigation is essential,
+requiring executing thousands or even millions of circuit variations to implement
+certain mitigation protocols. Transmitting these variations across cloud environments
```
Here we should distinguish between the estimator abstraction, which says nothing about mitigation, and our IBM hardware implementation---the primitives service---that has options to enable it.

ihincks commented on 0023-executor-primitive.md:
```diff
@@ -0,0 +1,188 @@
+# The proposal wants to acknowledge the transtionary reality in which the
+# primitives continue being the entry point for operating the quantum computer
+# and this is a low level interface for advanced usage.
+from qiskit.primitives.implementation import QuantumProgram, Executor
```
I expect that the `Executor` will need to come from `qiskit-ibm-runtime` instead: it needs to do things like POST to our service.

Instead, base class like BaseExecutor could live here.

I'm also wondering how this might all fit in with BackendV3, @ElePT , @jakelishman , @mtreinish : we might consider not having an `Executor` at all and reverting to something more like `backend.run()`. In this case we might need to have different semantics for sessions/batches than we have today.

ElePT replied:
> I expect that the Executor will need to come from qiskit-ibm-runtime instead: it needs to do things like POST to our service.

They need to, same as sampler and estimator.

> I'm also wondering how this might all fit in with BackendV3, @ElePT , @jakelishman , @mtreinish : we might consider not having an Executor at all and reverting to something more like backend.run(). In this case we might need to have different semantics for sessions/batches than we have today.

I think this requires understanding how this proposal affects the "primitives model" of execution we currently have. Will this replace the primitives? complement the primitives? I don't know how we could reasonably justify reverting to `backend.run()` if the `sampler` and `estimator` primitives continue existing...

delapuente replied:
Ok, I'm going to try to imitate the RFC for the [Estimator](https://github.com/Qiskit/RFCs/blob/master/0015-estimator-interface.md) in which the imports are not displayed, and elaborate on the concepts, then we can think of a strategy for adding this to Qiskit.

ihincks commented on 0023-executor-primitive.md:
```diff
@@ -0,0 +1,188 @@
+
+## Future Extensions
+
+The introduction of a low‑level sampler is only the first step. The long‑term
```
```suggestion
The introduction of a low‑level sampler (called an executor) is only the first step. The long‑term
```

delapuente committed:
Generalize the definition of the Estimator

Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>

delapuente committed:
Remove mention to a "low-level" primitive

delapuente committed:
Clarify samplex DSL description and its benefits

Refactor the description of the samplex DSL for clarity and consistency. Improve the explanation of its role in error mitigation and its benefits for users and developers.

delapuente committed:
Fix author list and update status in RFC 0023

delapuente committed:
Delve into the role of QuantumProgram and Executor within qiskit_ibm_runtime

delapuente committed:
Refactor documentation for ExecutorProgram usage

Delve into the key concepts of the Executor primitive, and its relation with samplex annotation and the samplex data structure.

delapuente committed:
Refine explanation on samplex generation decision

Clarify the reasoning behind the decision to avoid saving user-generated samplex in the client and emphasize the importance of local generation for inspectability and debuggability.

delapuente commented on 0023-low-level-sampler.md:
```diff
@@ -0,0 +1,188 @@
+# Prepare circuit for executing
+template, samplex = samplex.build()
+
+with Session(backend=backend) as session:
```
This is now addressed in its own section.

delapuente committed:
Update 0023-executor-primitive.md

Reframe to balance focus on the Executor primitive as a programming model, and introducing randomization semantics in Qiskit.

delapuente committed:
Update 0023-executor-primitive.md

Refine the demo and detailed design of the RFC.

SamFerracin commented on 0023-executor-primitive.md:
```diff
@@ -0,0 +1,335 @@
+	job = executor.run(noise_learner)
+  result = job.result()
+  
+  # Extrac noise map
+  noise_map = { layer: model for zip(circuit_layers, result }
```
```suggestion
  # Extract noise map
  noise_map = { layer: model for zip(circuit_layers, result) }
```

SamFerracin commented on 0023-executor-primitive.md:
```diff
@@ -0,0 +1,335 @@
+important for implementers to address in provider-specific documentation.
+
+## Alternative Approaches
+An alternative to introducing the **samplex** DAG is to generate all circuit
+variations entirely on the client side and transmit them to the backend for
+execution. This approach offers maximum flexibility and complete control over
+how variations are produced, since the backend would simply execute the
+provided circuits without influencing their structure. Users could implement
+any custom error mitigation workflow they wish, with no restrictions imposed by
+backend capabilities or vendor‑specific optimizations.
```
There is be another, more efficient alternative to "[generating] all circuit
variations entirely on the client side," which is to simply generate templates and parameters. However, a benefit of the `samplex` is that it allows generating parameters "on the fly:" vendors can set up their stack in such a way that their QPU can execute a set of parameters while a new set of parameters is being generated by the samplex. This can significantly speed up execution, with respect to both alternatives

delapuente replied:
Did not realize passing the samplex enabled lazy generation. Thank you.

delapuente committed:
Update 0023-executor-primitive.md

Co-authored-by: Samuele Ferracin <sam.ferracin@ibm.com>

delapuente committed:
Highlight lazy randomization expansion.
